File: capacitor-plugin/android/src/main/java/it/hubagency/spatialscan/RoomDataLoader.kt
package it.hubagency.spatialscan

import android.content.Context
import android.util.Log
import org.json.JSONObject
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Ricostruisce RoomExportData dai file JSON salvati sul device.
 * Utilizzato da SpatialScanPlugin.exportPdf() senza bisogno di una Activity.
 */
object RoomDataLoader {

    fun buildExportData(context: Context): RoomExportData? {
        val roomRecords = RoomHistoryManager.loadAll(context)
        if (roomRecords.isEmpty()) return null

        val rootId       = roomRecords.first().id
        val componentIds = CompositionGraph.getComponentRoomIds(context, rootId)
            .ifEmpty { roomRecords.map { it.id } }
        val roomMap      = roomRecords.associateBy { it.id }

        val allWalls    = mutableListOf<ExportWall>()
        val allPolygons = mutableListOf<Pair<String, List<Pair<Float, Float>>>>()

        for (roomId in componentIds) {
            val json   = RoomHistoryManager.loadRoomData(context, roomId) ?: continue
            val worldT = CompositionGraph.getTransform(context, roomId)
            val walls  = parseWalls(json)
            allWalls.addAll(if (worldT != null)
                transformWalls(walls, worldT.worldOffsetX, worldT.worldOffsetZ, worldT.worldRotRad)
            else walls)

            val name     = roomMap[roomId]?.name ?: roomId.take(6)
            val rawPoly  = parsePolygon(json)
            val worldPoly = if (worldT != null) {
                val c = cos(worldT.worldRotRad); val s = sin(worldT.worldRotRad)
                rawPoly.map { (x, z) ->
                    Pair(x * c - z * s + worldT.worldOffsetX, x * s + z * c + worldT.worldOffsetZ)
                }
            } else rawPoly
            allPolygons.add(name to worldPoly)
        }

        if (allWalls.isEmpty()) return null

        val totalArea = roomRecords.sumOf { it.area }
        val allX = allWalls.flatMap { listOf(it.startX.toDouble(), it.endX.toDouble()) }
        val allZ = allWalls.flatMap { listOf(it.startZ.toDouble(), it.endZ.toDouble()) }
        val dims = RoomDimensions(
            width     = (allX.max() - allX.min()).coerceAtLeast(0.0),
            length    = (allZ.max() - allZ.min()).coerceAtLeast(0.0),
            height    = allWalls.maxOf { it.height.toDouble() },
            area      = totalArea,
            perimeter = 0.0
        )
        return RoomExportData(allWalls, dims, allPolygons)
    }

    // ── Parsing ────────────────────────────────────────────────────────────────

    private fun parseWalls(json: JSONObject): List<ExportWall> = try {
        val walls = json.getJSONArray("walls")
        (0 until walls.length()).map { i ->
            val w  = walls.getJSONObject(i)
            val sp = w.getJSONObject("startPoint"); val ep = w.getJSONObject("endPoint")
            val sx = sp.optDouble("x", 0.0).toFloat(); val sz = sp.optDouble("z", 0.0).toFloat()
            val ex = ep.optDouble("x", 0.0).toFloat(); val ez = ep.optDouble("z", 0.0).toFloat()
            val dx = ex - sx; val dz = ez - sz
            val len = sqrt(dx * dx + dz * dz).coerceAtLeast(1e-6f)
            val h   = w.optDouble("height", 2.5).toFloat()
            val opsArr = w.optJSONArray("openings")
            val openings: List<ExportOpening> = if (opsArr != null) {
                (0 until opsArr.length()).mapNotNull { j ->
                    val o    = opsArr.optJSONObject(j) ?: return@mapNotNull null
                    val kind = runCatching { OpeningKind.valueOf(o.optString("kind", "DOOR")) }
                                  .getOrDefault(OpeningKind.DOOR)
                    ExportOpening(kind,
                        o.optDouble("offsetAlongWall", 0.0).toFloat(),
                        o.optDouble("width",  0.8).toFloat(),
                        o.optDouble("bottom", 0.0).toFloat(),
                        o.optDouble("height", 2.1).toFloat())
                }
            } else emptyList()
            ExportWall(id = w.optString("id", "w$i"),
                startX = sx, startZ = sz, endX = ex, endZ = ez,
                length = len, height = h,
                normalX = dz / len, normalZ = -(dx / len),
                dirX = dx / len, dirZ = dz / len,
                openings = openings)
        }
    } catch (e: Exception) {
        Log.e("RoomDataLoader", "parseWalls failed: ${e.message}"); emptyList()
    }

    private fun parsePolygon(json: JSONObject): List<Pair<Float, Float>> = try {
        val verts = json.getJSONObject("floor").getJSONArray("vertices")
        (0 until verts.length()).map { i ->
            val v = verts.getJSONObject(i)
            Pair(v.optDouble("x", 0.0).toFloat(), v.optDouble("z", 0.0).toFloat())
        }
    } catch (e: Exception) { emptyList() }

    private fun transformWalls(walls: List<ExportWall>, ox: Float, oz: Float, rot: Float): List<ExportWall> {
        val c = cos(rot); val s = sin(rot)
        return walls.map { w ->
            val sx = w.startX * c - w.startZ * s + ox; val sz = w.startX * s + w.startZ * c + oz
            val ex = w.endX   * c - w.endZ   * s + ox; val ez = w.endX   * s + w.endZ   * c + oz
            val dx = ex - sx; val dz = ez - sz
            val len = sqrt(dx * dx + dz * dz).coerceAtLeast(1e-6f)
            w.copy(startX = sx, startZ = sz, endX = ex, endZ = ez, length = len,
                   dirX = dx / len, dirZ = dz / len, normalX = dz / len, normalZ = -(dx / len))
        }
    }
}

File: ios/Plugin/RoomDataLoader.swift
import Foundation

/**
 * Ricostruisce RoomExportData dai file JSON salvati nel device.
 * Port 1:1 di RoomDataLoader.kt.
 *
 * Legge hub_rooms.json + hub_room_{id}.json, applica i world transforms
 * dal CompositionGraph e produce RoomExportData in world space.
 * Usato da SpatialScanPlugin.exportPdf() senza bisogno di un ViewController attivo.
 */
enum RoomDataLoader {

    // MARK: – Public entry points

    /// Ricostruisce RoomExportData usando `anchorRoomId` come radice della BFS.
    /// Usa sempre il componente connesso dell'ancora, non il primo record in assoluto.
    static func buildExportData(anchorRoomId: String) -> RoomExportData? {
        let roomRecords = RoomHistoryManager.shared.loadAll()
        guard !roomRecords.isEmpty else { return nil }

        let componentIds = CompositionGraph.shared.getComponentRoomIds(roomId: anchorRoomId)
        // Se nessun edge nel grafo ancora, esporta solo questa stanza
        let ids     = componentIds.isEmpty ? [anchorRoomId] : componentIds
        let roomMap = Dictionary(uniqueKeysWithValues: roomRecords.map { ($0.id, $0) })
        NSLog("[HUB_DIAG] RoomDataLoader anchorRoomId=%@ componentIds=[%@]",
              anchorRoomId, ids.joined(separator: ", "))

        var allWalls    = [ExportWall]()
        var allPolygons = [(String, [(Float, Float)])]()

        for roomId in ids {
            guard let json = RoomHistoryManager.shared.loadRoomData(id: roomId) else {
                NSLog("[HUB_DIAG] RoomDataLoader: no JSON for roomId=%@ — skipped", roomId)
                continue
            }
            let worldT = CompositionGraph.shared.getTransform(roomId: roomId)
            let walls  = parseWalls(json)
            if let wt = worldT {
                allWalls.append(contentsOf: transformWalls(walls,
                    ox: wt.worldOffsetX, oz: wt.worldOffsetZ, rot: wt.worldRotRad))
            } else {
                allWalls.append(contentsOf: walls)
            }

            let name     = roomMap[roomId]?.name ?? String(roomId.prefix(6))
            let rawPoly  = parsePolygon(json)
            let worldPoly: [(Float, Float)]
            if let wt = worldT {
                let c = cos(wt.worldRotRad); let s = sin(wt.worldRotRad)
                worldPoly = rawPoly.map { (x, z) in
                    (x * c - z * s + wt.worldOffsetX, x * s + z * c + wt.worldOffsetZ)
                }
            } else {
                worldPoly = rawPoly
            }
            allPolygons.append((name, worldPoly))
        }

        guard !allWalls.isEmpty else {
            NSLog("[HUB_DIAG] RoomDataLoader: no walls found for anchor=%@", anchorRoomId)
            return nil
        }

        let relevantRecords = ids.compactMap { roomMap[$0] }
        let totalArea = relevantRecords.reduce(0.0) { $0 + $1.area }
        let allX = allWalls.flatMap { [Double($0.startX), Double($0.endX)] }
        let allZ = allWalls.flatMap { [Double($0.startZ), Double($0.endZ)] }
        let dims = RoomDimensions(
            width:     max((allX.max() ?? 0) - (allX.min() ?? 0), 0),
            length:    max((allZ.max() ?? 0) - (allZ.min() ?? 0), 0),
            height:    Double(allWalls.map { $0.height }.max() ?? 2.5),
            area:      totalArea,
            perimeter: 0
        )
        NSLog("[HUB_DIAG] RoomDataLoader: built — walls=%d rooms=%d area=%.1f",
              allWalls.count, allPolygons.count, totalArea)
        return RoomExportData(walls: allWalls, dimensions: dims, roomPolygons: allPolygons)
    }

    /// Wrapper: usa il primo record salvato come ancora (chiamata da exportPdf senza contesto Composer).
    static func buildExportData() -> RoomExportData? {
        let roomRecords = RoomHistoryManager.shared.loadAll()
        guard !roomRecords.isEmpty else { return nil }
        return buildExportData(anchorRoomId: roomRecords.first!.id)
    }

    // MARK: – Parsing

    private static func parseWalls(_ json: [String: Any]) -> [ExportWall] {
        guard let wallsArr = json["walls"] as? [[String: Any]] else { return [] }
        return wallsArr.enumerated().compactMap { (i, w) in
            guard let sp = w["startPoint"] as? [String: Any],
                  let ep = w["endPoint"]   as? [String: Any] else { return nil }
            let sx = (sp["x"] as? NSNumber)?.floatValue ?? 0
            let sz = (sp["z"] as? NSNumber)?.floatValue ?? 0
            let ex = (ep["x"] as? NSNumber)?.floatValue ?? 0
            let ez = (ep["z"] as? NSNumber)?.floatValue ?? 0
            let dx = ex - sx; let dz = ez - sz
            let len = max(sqrt(dx*dx + dz*dz), 1e-6)
            let h   = (w["height"] as? NSNumber)?.floatValue ?? 2.5

            let openings: [ExportOpening]
            if let opsArr = w["openings"] as? [[String: Any]] {
                openings = opsArr.compactMap { o in
                    let kind = OpeningKind(rawValue: (o["kind"] as? String) ?? "") ?? .door
                    return ExportOpening(
                        kind:            kind,
                        offsetAlongWall: (o["offsetAlongWall"] as? NSNumber)?.floatValue ?? 0,
                        width:           (o["width"]           as? NSNumber)?.floatValue ?? 0.8,
                        bottom:          (o["bottom"]          as? NSNumber)?.floatValue ?? 0,
                        height:          (o["height"]          as? NSNumber)?.floatValue ?? 2.1
                    )
                }
            } else {
                openings = []
            }

            let wid = (w["id"] as? String) ?? "w\(i)"
            return ExportWall(
                id:      wid,
                startX:  sx, startZ: sz, endX: ex, endZ: ez,
                length:  len, height: h,
                normalX: dz / len, normalZ: -(dx / len),
                dirX:    dx / len, dirZ:     dz / len,
                openings: openings
            )
        }
    }

    private static func parsePolygon(_ json: [String: Any]) -> [(Float, Float)] {
        guard let floorDict = json["floor"] as? [String: Any],
              let verts = floorDict["vertices"] as? [[String: Any]] else { return [] }
        return verts.compactMap { v in
            guard let x = (v["x"] as? NSNumber)?.floatValue,
                  let z = (v["z"] as? NSNumber)?.floatValue else { return nil }
            return (x, z)
        }
    }

    private static func transformWalls(_ walls: [ExportWall],
                                       ox: Float, oz: Float, rot: Float) -> [ExportWall] {
        let c = cos(rot); let s = sin(rot)
        return walls.map { w in
            let sx = w.startX * c - w.startZ * s + ox
            let sz = w.startX * s + w.startZ * c + oz
            let ex = w.endX   * c - w.endZ   * s + ox
            let ez = w.endX   * s + w.endZ   * c + oz
            let dx = ex - sx; let dz = ez - sz
            let len = max(sqrt(dx*dx + dz*dz), 1e-6)
            return ExportWall(
                id:      w.id,
                startX:  sx, startZ: sz, endX: ex, endZ: ez,
                length:  len, height: w.height,
                normalX: dz / len, normalZ: -(dx / len),
                dirX:    dx / len, dirZ:     dz / len,
                openings: w.openings
            )
        }
    }
}
